### PR TITLE
Fixes #30538 - Update set of available events

### PR DESCRIPTION
**Pull request by ofedoren**
Adds CUD events for some models.

The change that requires attention is that default payload for events. Before only object id was returned, but now the whole object is being returned. This might solve reference errors in case of need to access e.g. `object.relation` for destroyed object. I've saved custom payload for other events, but CUD.

### CODE REVIEW

**Review by lzap — CHANGES_REQUESTED** [review]
LGTM, looks like you missed one tho:

```
app/models/model.rb
9:  set_hook :model_created, on: :create
10:  set_hook :model_updated, on: :update
11:  set_hook :model_destroyed, on: :destroy do |model|
```

**Comment by lzap** [comment]
To give you little bit of context @timogoebel we are dropping the idea of "almost zero payload" because this would make the webhooks plugins unnecessary complex. We want to use templating engine to render the `object` to its final form. Since the "user facing" API will be webhooks I think it's okay to "delay" the exposure of the Rails model and do this in webhook plugin.

**Comment by timogoebel** [comment]
> We want to use templating engine to render the `object` to its final form.

How can we make sure the API does not break here? What I mean is:
I, a proud Foreman user, download the cool ServiceNow glue script that uses Foreman's webhooks to talk to ServiceNow from Foreman's community webhook repository. Previously I changed all webhook templates to something weird, the community script therefore does not work for me at all. I'm frustrated and migrate all my workload to some cloud provider because they have a stable API.

**Comment by lzap** [comment]
I am not much following your hypothesis, however:

> How can we make sure the API does not break here?

By using our template engine, we are not giving up on stable API. It will simply not happen in this subscription API, which is quite low-level. We can make sure the templates (HTTP payload in webhooks) is stable - and we can easily achieve this with unit (render) tests by rendering those to prevent payload regressions.

**Comment by timogoebel** [comment]
> I am not much following your hypothesis

That's actually a good thing, I believe then this is not an issue.

> and we can easily achieve this with unit (render) tests by rendering those to prevent payload regressions.

Ok, so a user cannot change the templates? This is a developer-only thing? Isn't this the same as the rabl template from APIv2 then?

**Comment by ofedoren** [comment]
> LGTM, looks like you missed one tho:
> 
> ```
> app/models/model.rb
> 9:  set_hook :model_created, on: :create
> 10:  set_hook :model_updated, on: :update
> 11:  set_hook :model_destroyed, on: :destroy do |model|
> ```

@lzap, sorry, I don't not quite understand. Missed what exactly?.. For the `Model` model I've just refactored the old way of hooks setting.

> Ok, so a user cannot change the templates? This is a developer-only thing? Isn't this the same as the rabl template from APIv2 then?

@timogoebel, if I'm not mistaken, currently events and their payloads are used in foreman_webhooks only in terms of exposing information to a user. If this is not correct then the changes in this PR are quite destructive.

Regarding your recent questions:
 - Ideally, a user can change the templates to adjust the desired payload.
 - Not developer-only, since we are going to provide default templates which can be cloned and modified.
 - Pretty similar. I guess, the main difference is that a user can customize the result of template rendering.

I hope my comment helps a bit :)

**Comment by timogoebel** [comment]
> * Ideally, a user can change the templates to adjust the desired payload.

@ofedoren: Why would a user want to change the templates?
I think we should aim to provide a stable API here. I don't mean the contract between core and the plugin, we can adjust this at this early phase and have the actual payload definition in the plugin. What I mean is that third-party applications that want to use the webhook feature have to have a payload they can rely on to work.
What would happen if a user could change the Gitlab API as they want. Travis-CI would have a horrible time making their app work with every single repository. Right now this works because every Github repo in the entire internet has the same API and all the webhooks provide a stable, predictable payload.

Our initial idea was like this:
Foreman would just sent a minimal payload. If Foreman informs another application that e.g. a host was created, it would just pass the host id. The other application can than callback to Foreman and use e.g. APIv2 to query for more data because it knows the host id.
This works fine with Create and Update operations, but does not work with a Delete operation. If the other application would query Foreman for more information about host 123 that was just deleted, it would just get a 404 not found error.
That's why we added more information for the delete operation.

**Comment by lzap** [comment]
> Foreman would just sent a minimal payload.

Yes, I remember what we initially planed - to prevent users from changes in model layer. However it makes only sense when you have an external process being notified remotely with serialized payload (e.g. JSON in foreman_hooks plugin, my idea back then was Redis or some kind of messaging). The idea of fetching additional details via CLI/API makes sense in this context, but since we used ActiveSupport Notifications API to do the integration things suddenly change.

Now, even if we send just database ID in the payload, the plugin (e.g. webhooks) runs in the Ruby on Rails context (process), nothing can stop the plugin author from doing `Model.find(id)` and that's what all plugins would do from the day one. Therefore all the benefit of not exposing too much data is gone. And that's exactly what we were planning to do in webhooks during the template rendering - to (re)fetch the model from database.

From the original idea of actively blocking users from accessing unnecessary data it is just unnecessary step and the problem is still there. Thus, we have decided it would be best to give up on this feature and simply send the whole model object in the payload.

On top of that, it's so convenient to simply use the ActiveRecord model in the payload because we have the documented safe-mode (lists of methods, helpers and variables), users already have some know-how and they know what to expect. There is no need to have additional "fetch" helper in each template when it does not prevent plugin authors from doing the same. This only applies to webhooks plugin when we plan to use templates tho.

I understand this is quite departure from what we agreed on and I apologize from not including you in our discussion, we formed a small group (me and @ofedoren for the dev team) and discussed this with @ares and felt like this is the best option going forward.
In other words: instead of not sending whole model in the payload which does not prevent anything, we should probably encourage plugin authors not to expose whole model objects over an external APIs because these are subjects of change. But the reasonable way of doing that would be probably a sentence on our How to write a plugin wiki page.

**Comment by ares** [comment]
I think the key thing here is, we want Foreman to be able to own the payload definition. Users don't need to use any external script or application if they don't want to. With the payload defined as a template, we can construct whatever message the external system needs.

This does not go against the previous approach, where we'd only call external system with the id, an it would reach back for more information. We can still do that. But what this PR changes is, what is available in the payload renderer context. It's not only id but the whole AR object (its Jail more precisely). And the payload can be modified, it won't always be just `{id: $id}`

If we only pass id to the payload rendering context, whenever there was a need to build a custom payload with more information for the webhook, we'd have to `<% host = load_hosts(search: "id = #{@object_id}").first %>` or something similar. With the full record, payload can be easily constructed like in every other template. If the object id is available, nothing prevents user from accessing models (only safemode).

Like with any other template, we should lock the ones we seed by default. We should also use requirements in metadata.

@lzap will provide more context, if there's some follow up conversation on irc or elsewhere, please ping me too, I'd be interested to join

EDIT: he did while I was typing :-) perhaps I misunderstood the concern, but hopefully the intention is clearer

**Comment by ekohl** [comment]
The information @lzap provided is excellent to capture in a commit message. Capturing design decisions is what commit messages are for IMHO. It is also much easier for reviewers, especially those who weren't involved in some discussions.

**Comment by timogoebel** [comment]
Thanks, guys, I'm all for exposing the whole model in the internal API. My concern is just with rendering user-changeable templates in the webhook plugin. I don't see a world where this actually works.

Let me explain with yet another example.
If I want to send a slack message for every host that is created and slack would need two API requests (login and get a token, then send the actual message), I believe this won't work with the template engine.

**Comment by lzap** [comment]
> If I want to send a slack message for every host that is created and slack would need two API requests (login and get a token, then send the actual message), I believe this won't work with the template engine.

Correct and we should not be trying to fulfil this case. In this case, just use webhooks like you intented - there will be some kind of "Generic webhook template" which will deliver just id + hostname and it will be up to user to implement a service.

But for example Ansible (AWS) API allows to create inventory host entry in a single call, deleting is the same story. So it can work, why would we force users to write a custom service for that?

Let's talk real numbers, do you have a list of services which have publicly documented API and you would like to integrate Foreman with? I was trying to come up with something and all I got was Ansible AWX and ServiceNow (popular ticketing system) honestly. I can do a little research to find it how applicable rendering will be then.

**Comment by timogoebel** [comment]
> do you have a list of services which have publicly documented API and you would like to integrate Foreman with?

I would have added ServiceNow to the list as well. It's not something I'd personally want, but I've heard people asking for it. We are working on an integration with Netbox. That'll be a full plugin which we aim to open source soon.
I could see NetApp Filer or Backup products like Veeam or IBM's Spectrum Protect as well.

**Comment by ares** [comment]
@timogoebel I think your slack example would be still doable (without safe mode) but I would discourage from doing it this way. Instead, user should obtain the token manually and then use it in webhook definition (be it header, parameter, whatever). But in case it for whatever reason is more complicated, the original idea holds, you can still build your own service to handle it.

From my experience (mainly from home automation), if I have the power to customize the payload and url for a given event, I can wire in any external system. It would be PITA to write and maintain a service for each such system. What I like about the proposed behavior though, it does not add much complexity, while it allows both approaches.

**Comment by timogoebel** [comment]
> but I would discourage from doing it this way.

That's actually one concern I have with this approach. Users could try to add client API bindings to core so they can use them within templates. But fine, let's try this and see what users make of this feature.

How would users choose which template should be used for rendering the webhook payload? I hope this is going to be a per-hook setting.
What about storing secrets? Is this something we could design in a way that we could use it to allow hosts to retrieve secrets in the future?

**Comment by lzap** [comment]
> How would users choose which template should be used for rendering the webhook payload? I hope this is going to be a per-hook setting.

Yes, this is the plan.

> What about storing secrets? Is this something we could design in a way that we could use it to allow hosts to retrieve secrets in the future?

For now we would like to use encrypted field stored in Foreman DB as we do for other secrets for HTTP PLAIN auth. Agreed that some generic secret storage with a macro would be useful - this is currently out of scope.

**Comment by lzap** [comment]
I am going to merge this PR, started my research at https://community.theforeman.org/t/rfc-foreman-webhooks/19272/19

**Comment by lzap** [comment]
> @lzap, sorry, I don't not quite understand. Missed what exactly?.. For the `Model` model I've just refactored the old way of hooks setting.

I still see `set_hook` in model.rb filename, I think you have missed it. Use grep to find it.

**Comment by lzap** [comment]
Research: https://community.theforeman.org/t/rfc-foreman-webhooks/19272/19

**Review by lzap — APPROVED** [review]
Works fine. For the record, for testing at this stage you can create an initializer:

```
class DebugSubscriber < ::Foreman::BaseSubscriber
  def call(event)
    Rails.logger.debug("EVENT: #{event.inspect}")
  end
end

ActiveSupport::Notifications.subscribe(/.foreman$/, DebugSubscriber)
```

```
2020-08-26T13:28:21 [D|app|ded30073] EVENT: #<ActiveSupport::Notifications::Event:0x00007f6241ff2d78 @name="model_created.event.foreman", @payload={:object=>#<Model id: 3, name: "Test", info: "", created_at: "2020-08-26 11:28:21", updated_at: "2020-08-26 11:28:21", vendor_class: "", hardware_model: "">, :context=>{:remote_ip=>"::1", :request=>"ded30073-f3f3-4748-a309-e868f63dfc54", :session=>"3cecece3-994d-4443-ab9e-b59eb0010ee9", :user_login=>"admin", :user_admin=>true}}, @time=2020-08-26 13:28:21.508981287 +0200, @transaction_id="ec369a704fd3fd957cb2", @end=2020-08-26 13:28:21.508986637 +0200, @children=[], @cpu_time_start=0, @cpu_time_finish=0, @allocation_count_start=0, @allocation_count_finish=0>
```
